packages/ui/src/components/AppLayout/index.tsx

```diff
@@ -0,0 +1,171 @@
+    (
+        | {
+              navigationItems: SideNavigationProps.Item[];
+          }
+        | { navigation: ReactElement<SideNavigationProps> }
+    ) & { breadcrumbGroup?: ReactElement<BreadcrumbGroupProps> } & {
+        title: string;
+        defaultBreadcrumb?: string;
+    } & AppLayoutComponentProps;
+
+export interface AppLayoutContextApi {
+    appLayoutProps: AppLayoutComponentProps;
+    setAppLayoutProps: (props: AppLayoutComponentProps) => void;
+}
+
+const initialState = {
+    appLayoutProps: {},
+    setAppLayoutProps: () => {},
+};
+
+export const AppLayoutContext = createContext<AppLayoutContextApi>(initialState);
+
+const AppLayout: FC<PropsWithChildren<AppLayoutProps>> = ({
+    title,
+    defaultBreadcrumb = 'home',
+    children,
+    ...props
+}) => {
+    const navigate = useNavigate();
+    const [activeHref, setActiveHref] = useState('/');
+    const [activeBreadcrumbs, setActiveBreadcrumbs] = useState<BreadcrumbGroupProps.Item[]>([
+        { text: defaultBreadcrumb, href: '/' },
+    ]);
+    const [appLayoutProps, setAppLayoutProps] = useState<AppLayoutComponentProps>({});
+
+    const setAppLayoutPropsSafe = useCallback(
+        (props: AppLayoutComponentProps) => {
+            JSON.stringify(appLayoutProps) !== JSON.stringify(props) && setAppLayoutProps(props);
+        },
+        [appLayoutProps]
+    );
+
+    const onNavigate: CancelableEventHandler<BreadcrumbGroupProps.ClickDetail | SideNavigationProps.FollowDetail> =
+        useCallback(
+            (e) => {
+                if (!e.detail.external) {
+                    e.preventDefault();
+                    setAppLayoutProps({});
+                    setActiveHref(e.detail.href);
+
+                    const segments = [
+                        defaultBreadcrumb,
+                        ...e.detail.href.split('/').filter((segment) => segment !== ''),
+                    ];
+                    setActiveBreadcrumbs(
+                        segments.map((segment, i) => {
+                            const href = segments
+                                .slice(0, i + 1)
+                                .join('/')
+                                .replace('//', '/');
+                            return {
+                                href,
+                                text: segment,
+                            };
+                        })
+                    );
+                    navigate(e.detail.href);
+                }
+            },
+            [navigate, setAppLayoutProps, setActiveBreadcrumbs, defaultBreadcrumb]
+        );
+
+    return (
+        <>
+            {'header' in props ? (
+                props.header
+            ) : (
+                <NavHeader
+                    title={title}
+                    href={props.href}
+                    logo={props.logo}
+                    user={props.user}
+                    onSignout={props.onSignout}
+                />
+            )}
```

Comment:
cannot it be:
```js
{props.header ?? (
    <NavHeader
        title={title}
        href={props.href}
        logo={props.logo}
        user={props.user}
        onSignout={props.onSignout}
    />
)}
```

Reply:
Thanks. The header is a or type `NavHeaderProps | { header: ReactElement<TopNavigationProps>` .
`'header' in props` is used to cast the type to ReactElement<TopNavigationProps>`.